### PR TITLE
feat(resharding): state-stats improvements

### PR DESCRIPTION
**Pull request by wacban**
- group records by account id and aggregate it to get the total size of all records belonging to each account is used
- re-arrange records to be sorted by account id, rather than type and account id

results on mainnet
```
2023-11-07T13:25:00.063905Z  INFO state_viewer: s0.v1
2023-11-07T13:25:00.063968Z  INFO state_viewer: StateStats {
    total_size: 6.5 GB,
    total_count: 18507133,
    average_size: 351 B,
    middle_state_record: StateStatsStateRecord {
        account_id: "66fde23ac459779474beb6d1f8b72164e67946de.lockup.near",
        size: 344.9 KB,
    },
    top_accounts: [
        StateStatsStateRecord {
            account_id: "athlete.promotional.basketball.playible.near",
            size: 10.0 MB,
        },
        StateStatsStateRecord {
            account_id: "amm.counselor.near",
            size: 11.7 MB,
        },
        StateStatsStateRecord {
            account_id: "asset-manager.orderly-network.near",
            size: 142.2 MB,
        },
        StateStatsStateRecord {
            account_id: "anthropocene.mintbase1.near",
            size: 18.0 MB,
        },
        StateStatsStateRecord {
            account_id: "alpha-x.raddx.near",
            size: 405.3 MB,
        },
    ],
}


2023-11-07T13:29:24.136367Z  INFO state_viewer: s1.v1
2023-11-07T13:29:24.136422Z  INFO state_viewer: StateStats {
    total_size: 5.9 GB,
    total_count: 1,
    average_size: 5.9 GB,
    middle_state_record: StateStatsStateRecord {
        account_id: "aurora",
        size: 5.9 GB,
    },
    top_accounts: [
        StateStatsStateRecord {
            account_id: "aurora",
            size: 5.9 GB,
        },
    ],
}


2023-11-07T13:32:37.941254Z  INFO state_viewer: s2.v1
2023-11-07T13:32:37.941309Z  INFO state_viewer: StateStats {
    total_size: 5.9 GB,
    total_count: 10006107,
    average_size: 587 B,
    middle_state_record: StateStatsStateRecord {
        account_id: "deadmau5.mintbase1.near",
        size: 168.2 MB,
    },
    top_accounts: [
        StateStatsStateRecord {
            account_id: "canvas-war.feiyu.near",
            size: 50.5 MB,
        },
        StateStatsStateRecord {
            account_id: "jars.sweat",
            size: 77.9 MB,
        },
        StateStatsStateRecord {
            account_id: "fusotao.octopus-registry.near",
            size: 51.8 MB,
        },
        StateStatsStateRecord {
            account_id: "deadmau5.mintbase1.near",
            size: 168.2 MB,
        },
        StateStatsStateRecord {
            account_id: "bridge-validator.sputnik-dao.near",
            size: 134.4 MB,
        },
    ],
}

2023-11-07T13:40:05.003453Z  INFO state_viewer: s3.v1
2023-11-07T13:40:05.003511Z  INFO state_viewer: StateStats {
    total_size: 11.3 GB,
    total_count: 3113322,
    average_size: 3.6 KB,
    middle_state_record: StateStatsStateRecord {
        account_id: "tge-lockup.sweat",
        size: 3.5 GB,
    },
    top_accounts: [
        StateStatsStateRecord {
            account_id: "wallet.kaiching",
            size: 158.4 MB,
        },
        StateStatsStateRecord {
            account_id: "nft.nearapps.near",
            size: 321.6 MB,
        },
        StateStatsStateRecord {
            account_id: "x.paras.near",
            size: 605.6 MB,
        },
        StateStatsStateRecord {
            account_id: "token.sweat",
            size: 1.4 GB,
        },
        StateStatsStateRecord {
            account_id: "tge-lockup.sweat",
            size: 3.5 GB,
        },
    ],
}
```

### CODE REVIEW

File: tools/state-viewer/src/commands.rs [hunk 1]
```diff
@@ -1161,49 +1150,131 @@ pub(crate) fn print_state_stats(home_dir: &Path, store: Store, near_config: Near
...
+// Gets the flat state iterator and groups items by account id.
+fn get_state_stats_group_by<'a>(
+    chunk_view: &'a FlatStorageChunkView,
+    trie_storage: &'a TrieDBStorage,
+) -> GroupBy<
+    AccountId,
+    impl Iterator<Item = StateStatsStateRecord> + 'a,
+    impl FnMut(&StateStatsStateRecord) -> AccountId,
+> {
+    // let iter = chunk_view.iter_flat_state_entries(None, None);
```
Comment (shreyan-gupta) [inline tools/state-viewer/src/commands.rs]: nit: remove

File: tools/state-viewer/src/commands.rs [hunk 2]
```diff
@@ -1161,49 +1150,131 @@ pub(crate) fn print_state_stats(home_dir: &Path, store: Store, near_config: Near
...
-        tracing::info!(target: "state_viewer",  "{shard_uid:?}");
-        tracing::info!(target: "state_viewer",  "{state_stats:#?}");
+impl PartialOrd for StateStatsStateRecord {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
     }
 }
 
+// Gets the flat state iterator and groups items by account id.
+fn get_state_stats_group_by<'a>(
```
Comment (shreyan-gupta) [inline tools/state-viewer/src/commands.rs:1124]: Aah this is very clever! Took me some time to understand what's going on, had to read up on the kmerge and group_by functions.

File: tools/state-viewer/src/commands.rs [hunk 3]
```diff
@@ -1062,52 +1067,150 @@ pub(crate) fn clear_cache(store: Store) {
...
+    // account ids.
+    let iter = type_iters.kmerge().into_iter();
 
-        f.debug_struct("StateStatsStateRecord")
-            .field("account_id", &state_record_to_account_id(&state_record).as_str())
-            .field("type", &state_record.get_type_string())
-            .field("size", &self.size())
-            .finish()
+    // Finally, group by the account id.
+    iter.group_by(|state_record| state_record.account_id.clone())
```
Comment (shreyan-gupta) [inline tools/state-viewer/src/commands.rs:1169]: Would group_by `&state_record.account_id` work here?

**Review by shreyan-gupta — APPROVED** [review]
Could you please add the results from execution as well?

File: tools/state-viewer/src/commands.rs (hunk 2 above)
Comment (wacban) [reply tools/state-viewer/src/commands.rs:1124]: It took me real long to get the types, lifespans and whatnots in order but yeah it's pretty now :)

**Comment by wacban** [comment]
> Could you please add the results from execution as well?

Added to the PR description.

**Pull request merged**